pkgs/by-name/ru/rutabaga-gfx-ffi/0002-rutabaga_gfx-super-ugly-workaround-to-get-private-ke.patch

```diff
@@ -0,0 +1,75 @@
+From cfeedccee6c6e7c309369b5dc37b6525aac859b6 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?J=C3=B6rg=20Thalheim?= <joerg@thalheim.io>
+Date: Thu, 7 Dec 2023 18:45:37 +0100
+Subject: [PATCH 2/3] rutabaga_gfx: super ugly workaround to get private
+ keyboard resources forwarded
+MIME-Version: 1.0
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+
+Signed-off-by: Jörg Thalheim <joerg@thalheim.io>
+---
+ .../rutabaga_os/sys/linux/memory_mapping.rs   | 34 +++++++++++++++++--
+ 1 file changed, 32 insertions(+), 2 deletions(-)
+
+diff --git a/rutabaga_gfx/src/rutabaga_os/sys/linux/memory_mapping.rs b/rutabaga_gfx/src/rutabaga_os/sys/linux/memory_mapping.rs
+index 6d9f19124..afd8771c8 100644
+--- a/rutabaga_gfx/src/rutabaga_os/sys/linux/memory_mapping.rs
++++ b/rutabaga_gfx/src/rutabaga_os/sys/linux/memory_mapping.rs
+@@ -9,6 +9,7 @@
+ use nix::sys::mman::munmap;
+ use nix::sys::mman::MapFlags;
+ use nix::sys::mman::ProtFlags;
++use nix::errno::Errno;
+ 
+ use crate::rutabaga_os::descriptor::SafeDescriptor;
+ use crate::rutabaga_utils::RutabagaError;
+@@ -55,14 +56,43 @@ pub fn from_safe_descriptor(
+             // TODO(b/315870313): Add safety comment
+             #[allow(clippy::undocumented_unsafe_blocks)]
+             let addr = unsafe {
+-                mmap(
++                let mut res = mmap(
+                     None,
+                     non_zero_size,
+                     prot,
+                     MapFlags::MAP_SHARED,
+                     Some(descriptor),
+                     0,
+-                )?
++                );
++                if let Err(Errno::EPERM) = res {
++                    if prot == ProtFlags::PROT_READ {
```

Comment:
Why not just always check prot to get the right map flags, like [this](https://spectrum-os.org/git/spectrum/tree/pkgs/cloud-hypervisor/0002-virtio-devices-add-a-GPU-device.patch?id=a17f5d4abf9c4f49098cb65868f26c28d348ab0e#n239)?

Reply:
I don't see what check you are referring to.

Reply:
Why do you try mapping it first with MAP_SHARED, and only if that fails fall back to mapping with the flags based on prot?  Why not just always figure out the right flags up front?

Reply:
MAP_SHARED + PROT_READ only seem to fail for file descriptors that are based on memfd. Other types of file descriptors are fine as expected. Note that I actually create a copy of the content of the original memory, which is a rather ugly work around - however it seems to be only used for keyboard configuration...

Reply:
Do you have a nix expression to get all crosvm build dependencies btw?

Reply:
I usually just do `nix-build -A crosvm <nixpkgs>`.

Reply:
> MAP_SHARED + PROT_READ only seem to fail for file descriptors that are based on memfd. Other types of file descriptors are fine as expected.

Right, but there's still no harm doing MAP_PRIVATE in that case, right?  There's no need for two separate mmap calls.

Reply:
Ok. So you never ran `./tools/presubmit --no-delta all` when submitting changes to crosvm upstream?

Reply:
I have done but I don't remember how.  Doesn't it just run in podman?